tools/single_use/fix_double_escape.php

```diff
@@ -116,10 +119,70 @@
 
 //SECOND loop, depending on flags, report or fix values.
 if ($errorsDetected) {
-    //TODO CODE FOR CHECKING FIELD MAX LENGTH AND COMPARING TO ACTUAL LENGTH IN ORDER TO BUILD THE CONFIRMEDTRUNCATIONS array defined above
-
-    if ($useHistory) {
-        // TODO CODE FOR CHECKING HISTORY TABLE AND BUILDING BETTER REPAIR SUGGESTIONS
+	foreach ($escapedFields as $tableName => $fieldsArray) {
+		$fieldsList = "'" . implode($fieldsArray, "','") . "'";
+
+		$maxFieldLengths[$tableName] = $DB->pselectWithIndexKey(
+			"SELECT COLUMN_NAME, CHARACTER_MAXIMUM_LENGTH
+			FROM INFORMATION_SCHEMA.COLUMNS 
+			WHERE TABLE_NAME=:tbl AND COLUMN_NAME IN ($fieldsList)",
+			array('tbl' => $tableName),
+			'COLUMN_NAME'
+		);
+	}
+
+	// Start comparing the value length of the escaped entry to the character maximum for that field
+	// In order to begin identifying cases of truncation
+	foreach ($escapedEntries as $tableName => $rows) {
+		foreach ($rows as $cid => $entries) {
+			foreach ($entries as $field => $value) {
+				if (strlen($value) == $maxFieldLengths[$tableName][$field]['CHARACTER_MAXIMUM_LENGTH']) {
+					$confirmedTruncations[$tableName][$cid][$field] = $value;
+					$CIDs[] = $cid;
+					$allFields[] = $field;
+					$allTables[] = $tableName;
+				}
+			}
+		}
+	}
+
+if ($useHistory) {
+	// Construct history query
+	$listCIDs = "'" . implode($CIDs, "','") . "'";
+	$listFields = "'" . implode($allFields, "','") . "'";
+	$listTables = "'" . implode($allTables, "','") . "'";
+	$query = "SELECT primaryVals as CommentID, col, tbl, new, changeDate
+		FROM history 
+		WHERE primaryCols='CommentID' AND
+		primaryVals IN ($listCIDs) AND
+		tbl IN ($listTables) AND
+		col IN ($listFields)";	
```

Comment:
not sure this query would work, lets say 
 - commentID XYZ has an issue with instrument A.
 - commentID WXY has an issue with instrument B.

Then your results will include 
 - CommentId XYZ instrument B 
 - Commentid WXY instruemnt A

Both of which might not be problematic. I think this query needs to be exact and exhaustive.

Reply:
@jesscall

Reply:
Having the query be exact is very costly and I wasn't actually able to get results because the server kept disconnecting... 
This way, the query is not as costly and we do most of the heavy lifting in the script itself by comparing to the array of `confirmedTruncations` 

I know you mentioned it might be an issue with resourcing. Perhaps we can discuss offline and with IT and try to see if an exact query will work when more resources are placed on the history DB.